### PR TITLE
Add `Dataset.dtypes` property

### LINKED ISSUE #6714: `mypy` CI is failing on `main`
The most recent runs of the `mypy` CI on `main` are failing with:
```
xarray/core/dataset.py:6934: error: Incompatible types in assignment (expression has type "str", variable has type "Optional[Optional[Literal['Y', 'M', 'W', 'D', 'h', 'm', 's', 'ms', 'us', 'ns', 'ps', 'fs', 'as']]]")  [assignment]
```
I'm not sure what changed since the last pass, though.

### PR DESCRIPTION
**Pull request by headtr1ck**
- [x] Closes #6714
- [x] Tests added
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- [x] New functions/methods are listed in `api.rst`

Currently returns a Mapping from variable names to dtypes for ALL variables in the Dataset, including coordinates.
Possibly better to only return data_vars dtypes? Give me your thoughts on that, it is easy to change.

### CODE REVIEW

**Comment by max-sixty** [comment]
Great!

My inkling would be only data variables, given that's all we iterate over on a `Dataset`. We could also add this into the `DatasetCoords` / `DataArrayCoords` for completeness.

Any other thoughts?

**Comment by headtr1ck** [comment]
> My inkling would be only data variables, given that's all we iterate over on a `Dataset`. 

I would have thought the same. I have updated it now.

**Comment by dcherian** [comment]
See related discussion here: https://github.com/pydata/xarray/issues/1699

I like the idea of `.dtypes` only listing data variables and `coords.dtypes` for the rest. This would nicely generalize to `DataArray.coords.dtypes`

File: xarray/core/dataarray.py [hunk 1]
```diff
@@ -102,7 +102,7 @@
 
 def _infer_coords_and_dims(
     shape, coords, dims
-) -> tuple[dict[Any, Variable], tuple[Hashable, ...]]:
+) -> tuple[dict[Hashable, Variable], tuple[Hashable, ...]]:
```
[the remark this answers is not in the document]
Comment (headtr1ck) [reply xarray/core/dataarray.py:105]: Yes, this is typically an issue for input types but not for output types.

**Comment by headtr1ck** [comment]
Is it intentional that `DataArrayCoordinates` and `DatasetCoordinates` are not part of api?

**Comment by max-sixty** [comment]
> Is it intentional that `DataArrayCoordinates` and `DatasetCoordinates` are not part of api?

Good question, I'm not sure. They are not intended to be created by users IIUC, it's more like a "view" of the underlying coords. Does anyone have a view?

**Comment by TomNicholas** [comment]
> Good question, I'm not sure. They are not intended to be created by users IIUC, it's more like a "view" of the underlying coords. Does anyone have a view?

My understanding of these classes is that they are just slightly more complex mapping0like objects that we return instead of returning dicts in some cases. Sometimes they are intended to be immutable views and sometimes mutable IIUC. But I agree that they aren't intended to be instantiated directly by users.

**Comment by max-sixty** [comment]
Thank you v much @headtr1ck !

**Pull request merged**